### ISSUE TITLE
How to find the best layer in downstream task?

### ISSUE THREAD

**Issue by realgump** [facebookresearch/perception_models#37]
In the paper `Perception Encoder: The best visual embeddings are not at the output of the network`, Figure 13 presents the language alignment performance of features extracted from different layers.

<img width="770" alt="Image" src="https://github.com/user-attachments/assets/0813c02c-f1a3-450a-a478-74be11618290" />

For PElang-G, were 47 experiments (through 1M pretraining + 70M finetuning each) conducted to find out the best layer? What would be the recommended approach to find out the best layer when I am facing a new downstream task?

**Comment by dbolya** [comment]
We use the layer ablations as an analysis tool, not a way to select the best layer for downstream tasks. For PE lang, we didn't do any granular search or experimentation apart from what was in the ablation table (Table 9): 3 experiments--layer 50, layer 47, and layer 41. Using the second to last layer is common practice (Even the original Llava does it), and we chose to test the 4th to last layer because InternViT2.5 finds it helpful for bigger models.

The point of alignment, though, is so that you don't have to choose the best layer for your downstream task. If using the original core checkpoint, you would have to do a granular search to find the best layer, but with the best features aligned to the end (with PE lang and PE spatial), you don't have to do any searching and can just do the last layer.

Otherwise, if you don't want to use the pre-aligned models (or neither PE lang or PE spatial fits for your task), then you'll have to do a course to fine grid search on a held out validation set. For instance, you can check what DINOv2 does to evaluate on its downstream tasks. It's likely that future work could find a heuristic that helps select the optimal layer without having to search, but that's out of the scope of this work.
